File: wcp_sdk_typescript/src/rpc-client.ts
/**
 * JSON-RPC 2.0 client over WebSocket with exponential-backoff reconnect.
 */

export class WcpRpcError extends Error {
  constructor(public code: number, message: string, public data?: unknown) {
    super(message);
    this.name = "WcpRpcError";
  }
  isRetryable(): boolean {
    const d = this.data as { retry?: { retryable?: boolean } } | undefined;
    return !!d?.retry?.retryable;
  }
}

type Pending = {
  resolve: (v: unknown) => void;
  reject: (e: Error) => void;
};

export class RpcClient {
  private socket: WebSocket | null = null;
  private nextId = 1;
  private pending = new Map<number, Pending>();
  private queue: string[] = [];
  private reconnectAttempt = 0;
  private streamHandler?: (event: { event_type: string; payload: unknown }) => void;
  private closed = false;

  constructor(public readonly url: string) {}

  onStreamEvent(h: (event: { event_type: string; payload: unknown }) => void): void {
    this.streamHandler = h;
  }

  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      const WS: typeof WebSocket =
        (globalThis as unknown as { WebSocket: typeof WebSocket }).WebSocket;
      this.socket = new WS(this.url);
      this.socket.onopen = () => {
        this.reconnectAttempt = 0;
        this.flushQueue();
        resolve();
      };
      this.socket.onerror = (e) => reject(new Error("ws error: " + String(e)));
      this.socket.onmessage = (ev: MessageEvent) => this.handleMessage(ev);
      this.socket.onclose = () => {
        this.socket = null;
        if (!this.closed) this.scheduleReconnect();
      };
    });
  }

  private handleMessage(ev: MessageEvent): void {
    let parsed: unknown;
    try {
      parsed = JSON.parse(String(ev.data));
    } catch {
      return;
    }
    const obj = parsed as Record<string, unknown>;
    if ("event_type" in obj && this.streamHandler) {
      this.streamHandler({
        event_type: String(obj.event_type),
        payload: obj.payload,
      });
      return;
    }
    const id = obj.id as number | undefined;
    if (id === undefined) return;
    const p = this.pending.get(id);
    if (!p) return;
    this.pending.delete(id);
    if ("error" in obj) {
      const e = obj.error as { code: number; message: string; data?: unknown };
      p.reject(new WcpRpcError(e.code, e.message, e.data));
    } else {
      p.resolve(obj.result);
    }
  }

  private scheduleReconnect(): void {
    const delay = Math.min(
      30_000,
      500 * Math.pow(2, this.reconnectAttempt++) + Math.random() * 250,
    );
    setTimeout(() => {
      if (!this.closed) this.connect().catch(() => undefined);
    }, delay);
  }

  private flushQueue(): void {
    if (!this.socket || this.socket.readyState !== 1) return;
    while (this.queue.length > 0) {
      const msg = this.queue.shift()!;
      this.socket.send(msg);
    }
  }

  call<T = unknown>(method: string, params: unknown = {}): Promise<T> {
    const id = this.nextId++;
    const msg = JSON.stringify({ jsonrpc: "2.0", id, method, params });
    return new Promise<T>((resolve, reject) => {
      this.pending.set(id, {
        resolve: resolve as (v: unknown) => void,
        reject,
      });
      if (this.socket && this.socket.readyState === 1) {
        this.socket.send(msg);
      } else {
        if (this.queue.length >= 64) this.queue.shift();
        this.queue.push(msg);
      }
    });
  }

  close(): void {
    this.closed = true;
    try {
      this.socket?.close();
    } catch {
      // ignore
    }
    this.pending.clear();
  }
}

File: wcp_sdk_python/wcp_sdk/rpc_client.py
"""
JSON-RPC 2.0 client over WebSocket and HTTPS.

Async-first. WebSocket connection management with exponential-backoff reconnect.
"""
from __future__ import annotations

import asyncio
import json
import logging
import random
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Optional

import websockets

log = logging.getLogger("wcp_sdk.rpc_client")


class WcpRpcError(Exception):
    def __init__(self, code: int, message: str, data: Any = None) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.data = data

    def is_retryable(self) -> bool:
        if not isinstance(self.data, dict):
            return False
        retry = self.data.get("retry")
        if not isinstance(retry, dict):
            return False
        return bool(retry.get("retryable", False))


@dataclass(frozen=True)
class JsonRpcResponse:
    id: Any
    result: Any | None
    error: WcpRpcError | None


StreamHandler = Callable[[dict[str, Any]], Awaitable[None]]


class RpcClient:
    """Async JSON-RPC 2.0 client.

    Supports WebSocket transport with reconnect or HTTPS POST transport.
    """

    def __init__(self, url: str) -> None:
        self.url = url
        self._socket: Optional[websockets.WebSocketClientProtocol] = None
        self._next_id = 1
        self._pending: dict[int, asyncio.Future[Any]] = {}
        self._reader_task: Optional[asyncio.Task[None]] = None
        self._stream_handler: Optional[StreamHandler] = None
        self._closed = False
        self._connect_lock = asyncio.Lock()

    def on_stream_event(self, handler: StreamHandler) -> None:
        self._stream_handler = handler

    async def connect(self) -> None:
        async with self._connect_lock:
            if self._socket is not None:
                return
            attempt = 0
            while not self._closed:
                try:
                    self._socket = await websockets.connect(
                        self.url, ping_interval=20, ping_timeout=10
                    )
                    self._reader_task = asyncio.create_task(self._reader_loop())
                    return
                except Exception as exc:
                    log.warning("rpc connect failed attempt=%d: %s", attempt, exc)
                    attempt += 1
                    delay = min(30.0, 0.5 * (2**attempt) + random.random() * 0.25)
                    await asyncio.sleep(delay)

    async def _reader_loop(self) -> None:
        assert self._socket is not None
        try:
            async for raw in self._socket:
                try:
                    data = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                if isinstance(data, dict) and "event_type" in data:
                    if self._stream_handler is not None:
                        await self._stream_handler(data)
                    continue
                req_id = data.get("id")
                fut = self._pending.pop(int(req_id), None) if req_id is not None else None
                if fut is None:
                    continue
                if "error" in data:
                    err = data["error"]
                    fut.set_exception(
                        WcpRpcError(
                            err.get("code", -32603),
                            err.get("message", ""),
                            err.get("data"),
                        )
                    )
                else:
                    fut.set_result(data.get("result"))
        except websockets.ConnectionClosed:
            pass

    async def call(
        self, method: str, params: Optional[dict[str, Any]] = None
    ) -> Any:
        if self._socket is None:
            await self.connect()
        assert self._socket is not None
        req_id = self._next_id
        self._next_id += 1
        fut: asyncio.Future[Any] = asyncio.get_event_loop().create_future()
        self._pending[req_id] = fut
        await self._socket.send(
            json.dumps(
                {
                    "jsonrpc": "2.0",
                    "id": req_id,
                    "method": method,
                    "params": params or {},
                }
            )
        )
        return await fut

    async def close(self) -> None:
        self._closed = True
        if self._reader_task is not None:
            self._reader_task.cancel()
        if self._socket is not None:
            await self._socket.close()
        self._socket = None
